ocaml/xapi/xapi_host.ml

```diff
@@ -2923,6 +2923,74 @@ let emergency_reenable_tls_verification ~__context =
   Helpers.touch_file Constants.verify_certificates_path ;
   Db.Host.set_tls_verification_enabled ~__context ~self ~value:true
 
+(** Issue an alert if /proc/sys/kernel/tainted indicates particular kernel
+    errors. Will send only one alert per reboot *)
+let alert_if_kernel_broken =
+  let __context = Context.make "host_kernel_error_alert_startup_check" in
+  (* Check all the alerts since last reboot. Only done once at toolstack
+     startup, we track if alerts have been issued afterwards internally *)
+  let self = Helpers.get_localhost ~__context in
+  let boot_time =
+    Db.Host.get_other_config ~__context ~self
+    |> List.assoc "boot_time"
+    |> float_of_string
+  in
+  let all_alerts =
+    [
+      (* processor reported a Machine Check Exception (MCE) *)
+      (4, Api_messages.kernel_is_broken "MCE")
+    ; (* bad page referenced or some unexpected page flags *)
+      (5, Api_messages.kernel_is_broken "BAD_PAGE")
+    ; (* kernel died recently, i.e. there was an OOPS or BUG *)
+      (7, Api_messages.kernel_is_broken "BUG")
+    ; (* kernel issued warning *)
+      (9, Api_messages.kernel_is_broken_warning "WARN")
+    ; (* soft lockup occurred *)
+      (14, Api_messages.kernel_is_broken_warning "SOFT_LOCKUP")
+    ]
+  in
+  (* Only add an alert if
+     (a) an alert wasn't already issued for the currently booted kernel *)
+  let possible_alerts =
+    ref
+      (all_alerts
+      |> List.filter (fun (_, alert_message) ->
+             let alert_already_issued_for_this_boot =
+               Helpers.call_api_functions ~__context (fun rpc session_id ->
+                   Client.Client.Message.get_all_records ~rpc ~session_id
```

Comment:
Performing this call while loading the module seems dangerous, it looks like it might cause a dependency loop.

Does it block until the socket is available? Does the socket depend on this module before it's available? (message forwarding depends on this module)

Using a lazy value, and only filling it up on the first call ro ma n independent thread seems safer.

Reply:
This is much more sensible, done, thank you! (likely the tests were failing for this exact reason)

Reply:
Lazy is also not thread safe (and crashed on older OCaml versions), so we have to be very careful how to use it. Another way to avoid lazy is to add a new function to xapi's startup sequence in xapi.ml, or to the periodic scheduler init file.

We probably want to check this value periodically anyway, not just once on startup (because errors can happen later during a host's lifecycle, it is very unlikely for something to have happened early on boot).